### ISSUE TITLE
core audio output error: no suitable audio output module

### ISSUE THREAD

**Issue by steve-donovan** [caprica/vlcj#448]
INFO  vlcj: 3.10.0
INFO  java: 1.8.0_66 Oracle Corporation
INFO  os: Mac OS X 10.11.2 x86_64
INFO  vlc: 2.2.1 Terry Pratchett (Weatherwax), changeset 2.2.1-0-ga425c42
jna-4.1.0
jna-platform-4.1.0

Have JavaFX app and therefore tried implementing the DirectMediaPlayer example you kindly provided in another project. Kept getting a seg crash so thought I'd try updating to above libs etc. Now get following errors instead:

...
[00007faa8e831d18] core audio output error: no suitable audio output module
...
[00007faa8c4c4c88] core input error: open of `file:///Users/daddio/Desktop/EB-Media/EB_Music_Mix_1/test.mp3' failed
[00007faa8c4c4c88] core input error: Your input can't be opened
[00007faa8c4c4c88] core input error: VLC is unable to open the MRL 'file:///Users/daddio/Desktop/Media/Music_Mix_1/test.mp3'. Check the log for details.
...

I'm puling my hair out - by trying to solve one problem I appear to create new ones - 1 step forward, 2 steps back. Hope you can advise. Likely I've done something silly, but i hacked example so hope not. My player code is :

```
package com.digigames;

import java.nio.ByteBuffer;
import java.util.logging.Logger;

import com.sun.jna.Memory;

import javafx.animation.AnimationTimer;
import javafx.scene.canvas.Canvas;
import javafx.scene.image.PixelFormat;
import javafx.scene.image.PixelWriter;
import javafx.scene.image.WritablePixelFormat;
import uk.co.caprica.vlcj.component.DirectMediaPlayerComponent;
import uk.co.caprica.vlcj.player.MediaPlayer;
import uk.co.caprica.vlcj.player.MediaPlayerEventAdapter;
import uk.co.caprica.vlcj.player.MediaPlayerFactory;
import uk.co.caprica.vlcj.player.direct.BufferFormat;
import uk.co.caprica.vlcj.player.direct.BufferFormatCallback;
import uk.co.caprica.vlcj.player.direct.DefaultDirectMediaPlayer;
import uk.co.caprica.vlcj.player.direct.DirectMediaPlayer;
import uk.co.caprica.vlcj.player.direct.format.RV32BufferFormat;

public class DirectMediaPlayerCanvas {
	
    private static final Logger LOG = Logger.getLogger(ExtremeBingo.class.getName());

    private final Canvas canvas;
	private final PixelWriter pixelWriter;
	private final WritablePixelFormat<ByteBuffer> pixelFormat;
	
	private final DirectMediaPlayerComponent mediaPlayerComponent;
	
	private final AnimationTimer timer;
	
	private String mediaPlayerName;
	
	public DirectMediaPlayerCanvas(String mediaPlayerName) {
		this();
		this.mediaPlayerName = mediaPlayerName;
	}
	
	public DirectMediaPlayerCanvas() {
		canvas = new Canvas();
		pixelWriter = canvas.getGraphicsContext2D().getPixelWriter();
		pixelFormat = PixelFormat.getByteBgraInstance();
		mediaPlayerComponent = new MediaPlayerComponent();
		mediaPlayerComponent.getMediaPlayer().addMediaPlayerEventListener(new MediaPlayerEventAdapter() {
			
			@Override
			public void playing(MediaPlayer mediaPlayer) {
				LOG.info("MEDIA (playing) " + mediaPlayerName);
				if (timer != null) {
					timer.stop();
				}
				timer.start();
			}
			
			@Override
			public void stopped(MediaPlayer mediaPlayer) {
				LOG.info("MEDIA (stopped) " + mediaPlayerName);
				timer.stop();
			}

			@Override
			public void finished(MediaPlayer mediaPlayer) {
				LOG.info("MEDIA (finished) " + mediaPlayerName);
				timer.stop();
			}        
			
			@Override
			public void paused(MediaPlayer mediaPlayer) {
				LOG.info("MEDIA (paused) " + mediaPlayerName);
				timer.stop();
			}        

			@Override
			public void error(MediaPlayer mediaPlayer) {
				LOG.info("MEDIA (error) " + mediaPlayerName);
				timer.stop();
			}        
		});
		
		timer = new AnimationTimer() {
			@Override
			public void handle(long now) {
				renderFrame();
			}
		};
	}
	
	public Canvas getCanvas() {
		return canvas;
	}
	
    private class MediaPlayerComponent extends DirectMediaPlayerComponent {
        public MediaPlayerComponent() {
            super(new CanvasBufferFormatCallback());
        }
    }

    private class CanvasBufferFormatCallback implements BufferFormatCallback {
        @Override
        public BufferFormat getBufferFormat(int sourceWidth, int sourceHeight) {
        	final int canvasWidth = (int)canvas.getWidth();
        	final int canvasHeight = (int)canvas.getHeight();
        	
        	LOG.info(String.format("%s souceWidth [%d] sourceHeight [%d]", mediaPlayerName, sourceWidth, sourceHeight));
        	LOG.info(String.format("%s canvasWidth [%d] canvasHeight [%d]", mediaPlayerName, canvasWidth, canvasHeight));
        	
            return new RV32BufferFormat(canvasWidth, canvasHeight);
        }
    }

    // convenience method
    public DirectMediaPlayer getMediaPlayer() {
    	return mediaPlayerComponent.getMediaPlayer();
    }

    // convenience method
    public MediaPlayerFactory getMediaPlayerFactory() {
    	return mediaPlayerComponent.getMediaPlayerFactory();
    }

    // convenience method
    public void release() {
    	mediaPlayerComponent.release(true);
    }
    
    private final void renderFrame() {
        Memory[] nativeBuffers = mediaPlayerComponent.getMediaPlayer().lock();
        if (nativeBuffers != null) {
            // FIXME there may be more efficient ways to do this...
            // Since this is now being called by a specific rendering time, independent of the native video callbacks being
            // invoked, some more defensive conditional checks are needed
            Memory nativeBuffer = nativeBuffers[0];
            if (nativeBuffer != null) {
                ByteBuffer byteBuffer = nativeBuffer.getByteBuffer(0, nativeBuffer.size());
                BufferFormat bufferFormat = ((DefaultDirectMediaPlayer) mediaPlayerComponent.getMediaPlayer()).getBufferFormat();
                if (bufferFormat.getWidth() > 0 && bufferFormat.getHeight() > 0) {
                    pixelWriter.setPixels(0, 0, bufferFormat.getWidth(), bufferFormat.getHeight(), pixelFormat, byteBuffer, bufferFormat.getPitches()[0]);
                }
            }
        }
        mediaPlayerComponent.getMediaPlayer().unlock();
    }
}
```

called like : 

```
        DirectMediaPlayerCanvas audienceMediaPlayer = new DirectMediaPlayerCanvas("audienceMediaPlayer");
        audienceMediaPlayer.getCanvas().setWidth((int) audienceAVQuizDisplayAreaVideo.getWidth());
        audienceMediaPlayer.getCanvas().setHeight((int) audienceAVQuizDisplayAreaVideo.getHeight());
        audienceAVQuizDisplayAreaVideo.getChildren().add(audienceMediaPlayer.getCanvas());

            if (audienceMediaPlayer.getMediaPlayer().startMedia(NEXBG)) {
            	audienceMediaPlayer.getMediaPlayer().setRate(Float.parseFloat("1.00"));
            	audienceMediaPlayer.getMediaPlayer().setVolume((int) slider_volume_controll.getValue());
            }
            else {
            	LOG.severe(String.format("Media Item Failed to Play [%s]", NEXBG));
            }

```

**Comment by steve-donovan** [comment]
ok so tried : 
vlcj 3.10.0 with jna 4.2.1
vlcj 3.10.0 with jna 4.1.0
vlcj 3.9.0 with jna 3.5.2
vlcj 3.8.0 with jna 3.5.2

will not try reverting VLC 2.2.1 a version at a time

**Comment by steve-donovan** [comment]
playing one particular video file with startMedia, and then using startMedia on same MediaPlayer to play different video:

JNA: Callback uk.co.caprica.vlcj.player.direct.DefaultDirectMediaPlayer$LockCallback@4d7d3f8e threw the following exception:
java.lang.NullPointerException
	at uk.co.caprica.vlcj.player.direct.DefaultDirectMediaPlayer$LockCallback.lock(DefaultDirectMediaPlayer.java:263)
	at sun.reflect.GeneratedMethodAccessor45.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at com.sun.jna.CallbackReference$DefaultCallbackProxy.invokeCallback(CallbackReference.java:470)
	at com.sun.jna.CallbackReference$DefaultCallbackProxy.callback(CallbackReference.java:500)
java(4325,0x700001e25000) malloc: *** error for object 0x7fa2f0ac8e00: pointer being freed was not allocated
*** set a breakpoint in malloc_error_break to debug

**Comment by steve-donovan** [comment]
i noticed for the particular file causing this issue, getBufferFormat is called twice when on others its only called once. the video plays wonderfully, the crash is ONLY when moving to next movie

**Comment by caprica** [comment]
1. Keep swapping versions of things is most likely to be futile. I can only recommend latest official released VLC, latest official release vlcj, and JNA 4.1.0.

2. For a local file, why are you using a URL? i.e. the "file://"? Just use the local filename.

3. The NPE you posted is here:

            planes.getPointer().write(0, nativeBuffers, 0, nativeBuffers.length);

So planes is null, or nativeBuffers is null.

That shouldn't happen of course.

It's pretty difficult to say anything more about this, frankly.

What happens if you try the vlcj/vlcj-javafx examples?

4. Maybe post a self-contained, single-class, compilable, executable example that demonstrates your problem.

**Comment by steve-donovan** [comment]
Really appreciate the quick reply. Regressing to VLC 2.1.5 fixed the file name issue - I wasnt actually passing a URL, just a filepath, for some reason it was being converted to one. Something has changed in VLC since 2.1.5.

I'll try the exmaple you mentioned shortly and let you know outcome

**Comment by caprica** [comment]
But regressing to VLC 2.1.5 is hardly a fix!

**Comment by steve-donovan** [comment]
haha very true but my filename didnt change so VLC introduced a change that broke something somewhere

**Comment by steve-donovan** [comment]
Hi

used JavaFXDirectRenderingTest & AnimationTimerJavaFXDirectRenderingTest examples, only provided a file name for the FIXME

SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
[0x7ffb3bc73800] main vout display error: Failed to change zoom
[0x7ffb3bc7fef0] main vout display error: Failed to change zoom
[0x7ffb3bc73800] main vout display error: Failed to set on top
[0x7ffb3bc7fef0] main vout display error: Failed to set on top
[0x7ffb3bc73800] main vout display error: Failed to change source AR
[0x7ffb3bc7fef0] main vout display error: Failed to change source AR
JNA: Callback uk.co.caprica.vlcj.player.direct.DefaultDirectMediaPlayer$LockCallback@e87c563 threw the following exception:
java.lang.NullPointerException
	at uk.co.caprica.vlcj.player.direct.DefaultDirectMediaPlayer$LockCallback.lock(DefaultDirectMediaPlayer.java:263)
	at sun.reflect.GeneratedMethodAccessor4.invoke(Unknown Source)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:497)
	at com.sun.jna.CallbackReference$DefaultCallbackProxy.invokeCallback(CallbackReference.java:455)
	at com.sun.jna.CallbackReference$DefaultCallbackProxy.callback(CallbackReference.java:485)
objc[4818]: Method cache corrupted. This may be a message to an invalid object, or a memory error somewhere else.
objc[4818]: receiver 0x7ffb3971d7e0, SEL 0x7fff9cafb2d9, isa 0x7fff7adb6a78, cache 0x7fff7adb6a88, buckets 0x7ffb3b236000, mask 0x3f, occupied 0x15
objc[4818]: receiver 48 bytes, buckets 1536 bytes
objc[4818]: selector 'objectForKey:'
objc[4818]: isa '__NSDictionaryM'
objc[4818]: Method cache corrupted.
#
# A fatal error has been detected by the Java Runtime Environment:
#
#  SIGILL (0x4) at pc=0x00000001274e678e, pid=4818, tid=63747
#
[thread 24323 also had an error]
# JRE version: Java(TM) SE Runtime Environment (8.0_66-b17) (build 1.8.0_66-b17)
[thread 55299 also had an error]
# Java VM: Java HotSpot(TM) 64-Bit Server VM (25.66-b17 mixed mode bsd-amd64 compressed oops)
# Problematic frame:
# C  [libvlccore.7.dylib+0x4078e]  ThreadDisplayPicture+0x71e
#
# Failed to write core dump. Core dumps have been disabled. To enable core dumping, try "ulimit -c unlimited" before starting Java again
#
# An error report file with more information is saved as:
# /Users/daddio/Documents/vlcj_testing/Test/hs_err_pid4818.log
#
# If you would like to submit a bug report, please visit:
#   http://bugreport.java.com/bugreport/crash.jsp
#

**Comment by caprica** [comment]
Are you saying it's OK until it reaches the end of the video?

**Comment by steve-donovan** [comment]
Yes. It plays perfectly until either it's natural end, or if you end it to play another video.

**Comment by caprica** [comment]
The vlcj-javafx examples play just fine for me, when the video stops, and when auto-repeating the video. 

I see no such crashes, no NPE, it just works.

**Comment by steve-donovan** [comment]
I have success with most videos too. Some just seem different. The working ones call getBufferFormat once, they ones that crash call idly twice for from reason. 
Can I send you a video that breaks you example code? It's 120mb so I can't post on here.

**Comment by steve-donovan** [comment]
Also I'm on OSX as opposed to Windows.

**Comment by caprica** [comment]
Can you share a link via Dropbox/GoogleDrive etc? Don't post your link here, email me direct.

**Comment by steve-donovan** [comment]
Sure thing

**Comment by steve-donovan** [comment]
can’t for life of me find your email address

**Comment by caprica** [comment]
http://capricasoftware.co.uk/#/contact

**Comment by caprica** [comment]
So using MediaInfo shows there are two video tracks (and in fact two audio tracks) in that mp4 file.

If you play your file in vlc, it plays one track in the main application window and at the same time it opens another window to play the second track.

This might be a bug in vlc with how it handles media with multiple video tracks.

This fits with your description of the buffer format callback being called twice.

There is nothing in LibVLC to support playing multiple video tracks at the same time.

This is not something that can be fixed in vlcj.

**Comment by steve-donovan** [comment]
How strange, I’ve never come across an mp4 with dual tracks. I really appreciate your time on this Mark, I’m sorry it wasn’t something more interesting.

I did note that since VLC 2.1.5 they’ve changed the lib and plugin folder layout, and that silly filename error. I’d like to upgrade to latest of everything, but couldn’t even find workaround for the filename issue.

**Comment by caprica** [comment]
I don't understand your filename error, I use the latest VLC/vlcj and I don't see any such problem.

**Comment by steve-donovan** [comment]
on mac osx?

**Comment by caprica** [comment]
I do test on OSX, but have not done so recently. I can see no reason why this would behave any differently on OSX.

**Comment by steve-donovan** [comment]
so strange. it was fine but since 2.1.5 the issue hit, and i tried 2.2.1 and 2.2.2, same issue. hate silly little things like this. i have my fingers crossed someone else will hit the issue and post something up with a solution/workaround

**Comment by caprica** [comment]
Just tested a local file on OSX with VLC 2.2.0 and vlcj 3.10.0.

Works just fine.

You should post what you use to call mediaPlayer.playMedia(filename) - i.e. what is "filename".

**Comment by steve-donovan** [comment]
ok will wind everything forward again and give it a try. Not sure whether makes any difference, but I call startMedia instead of playMedia

**Comment by caprica** [comment]
startMedia and playMedia both call prepareMedia, that is where the filename is detected as either a local file (path) or a "location" in VLC vernacular.

**Comment by steve-donovan** [comment]
surely point is I don’t change the name of file between VLC versions, I simply change the libs and re-run exact same code…. hmmm that makes it even more of a mystery.
Would the JNA/Platofrm versions make any difference? are you using 4.2.1?

**Comment by caprica** [comment]
LibVLC changed the way libvlc_media_new_location and libvlc_media_new_path work, so that might be the cause.

JNA/Platform is irrelevant here.

vlcj is only supported with JNA 4.1.0 at the moment, the 4.2.x series has a regression that should be fixed in the next release of JNA, whenever that happens.

I'm going to close this issue now as it's basically done.

If you do have a genuine filename error, just go ahead and open a new issue.

**Comment by steve-donovan** [comment]
Thanks very much again Mark. I’ll have a play with the filename thing and see if i can generate better clues
